### ISSUE TITLE
Inaccurate country boundaries causing incorrect outputs

### ISSUE THREAD

**Issue by JimBacon** [BiologicalRecordsCentre/ABLE#631]
The country boundaries that Indicia uses to determine in which country a record has been made are polygons. Innaccuracies in the polygons give rise to two sorts of errors when records are close to the boundaries.

- The record may appear to be in the wrong country or no country meaning it is omitted from outputs.
- The record may appear to be in more than one country meaning it can occur multiple times in outputs.

Making the boundaries more accurate means increasing the level of detail. This will make the boundaries more complex and increase the time required to process them. This could be mitigated by a two-stage process where more accurate polygons are used when a record is in close proximity to the boundary. Even with perfect boundaries there is still going to be a dilemma when the accuracy of the recorded location is insufficient to place it wholly in one country.

It might be more reliable to determine the country from user input.

**Comment by johnvanbreda** [comment]
> This could be mitigated by a two-stage process where more accurate polygons are used when a record is in close proximity to the boundary.

Just to note that if the geometry intersections are processed by PostGIS using an index, then it's not clear that a 2 step process will help very much. The GIST index will be doing a fast bounding box query, then an exact check only where there is a hit.

Ideally we should do a bit of testing to compare the times of the following approaches:
1) Find intersections between a quantity of records against a quantity of location boundaries (high precision).
2) Find intersections between a quantity of records against a quantity of location boundaries (low precision).
3) Find intersections between a quantity of records against a quantity of location boundaries (low precision), for any hits, confirm them against a high precision version of the boundary.

**Comment by JimBacon** [comment]
Also worth noting that, while raised here because it is affecting the EBMS results, the issue is relevant to many more projects.

**Comment by JimBacon** [comment]
For info, boundaries can be obtained in various resolutions from https://ec.europa.eu/eurostat/web/gisco/geodata/reference-data/administrative-units-statistical-units/countries.

From inspection, the warehouse has the 1:10 million scale maps loaded. For, e.g. the island of Lanzarote, it looks as follows.
![image](https://github.com/BiologicalRecordsCentre/ABLE/assets/2545390/b2018b4c-eb31-4a39-8b92-1fbb37f26da5)

A comparison of the 1:10 million and the 1:1 million scale looks as follows (red 1:10 million, purple 1:1 million):
![image](https://github.com/BiologicalRecordsCentre/ABLE/assets/2545390/25861868-de98-40eb-b440-9da6869b35e5)

While the 1:1 million is clearly vastly more accurate, it will still exhibit the same problems we are having, just to a much lesser extent.

**Comment by CrisSevilleja** [comment]
Could not be possible to make a buffer around the boundaries to include the whole territory? especially for islands where you don't have an issue of including other countries/territories.

**Comment by JimBacon** [comment]
Yes, something like that is possible. I feel as if someone did something for Sweden to help incorporate its many small islands. Maybe there is already a boundary which includes territorial waters.

**Comment by JimBacon** [comment]
The case where a record appears to be in multiple countries is not down to the inaccuracy of the boundaries. Certainly I have seen vice county boundaries which overlap but the Countries 2016 boundaries, on cursory inspection, do not.

Instead, what I had not appreciated is that the spatial indexing is done on the public_sref and the record in question was marked sensitive and blurred to 10km. The result is that a point in Switzerland is blurred to a polygon overlapping France and Germany as well. See illustration.

![image](https://github.com/BiologicalRecordsCentre/ABLE/assets/2545390/a4983693-e6fd-495b-b481-394e9526e6c1)

The issue of the record being too inaccurate to locate to a single country was already mentioned.

**Comment by JimBacon** [comment]
> Yes, something like that is possible. I feel as if someone did something for Sweden to help incorporate its many small islands. Maybe there is already a boundary which includes territorial waters.

Perhaps something useful at https://www.marineregions.org/downloads.php#marbound
or https://github.com/tdwg/wgsrpd

**Comment by johnvanbreda** [comment]
@JimBacon you are correct about Sweden - I have found an email from September 2021 in which I said:
I’ve just replaced the Sweden boundary loaded on the warehouse by doing the following:
1.	Downloading the 1m precision (best available) from the Eurostat NUTS layers.
2.	Buffering it by 1000m.
3.	Clipping off any of the buffered area which overlaps another country polygon according to the NUTS layer (e.g. Norwegian border). So, the buffer only applies out to sea.

This was for marine buffering though, not land-borders.

**Comment by JimBacon** [comment]
> Clipping off any of the buffered area which overlaps another country polygon according to the NUTS layer (e.g. Norwegian border). So, the buffer only applies out to sea.

This is great when only applied to one country but I can see it being a problem when applied to multiple adjacent countries. There is a solution documented at https://gis.stackexchange.com/questions/264416/splitting-overlap-between-polygons-and-assign-to-nearest-polygon-using-postgis but it would be a big task to apply globally.

**Comment by JimBacon** [comment]
As a test I have 

1. Downloaded the Union of the ESRI Country shapefile and the Exclusive Economic Zones (version 3) from https://www.marineregions.org/downloads.php#unioneezcountry
2. Used QGIS to merge the 8 Spanish territories and exported this as a .shp file
3. Made a backup table of the existing Countries 2016 locations on the warehouse in a table called `locations_backup_temp`
4. Uploaded the new .shp file using the warehouse website which updated the location boundary.
5. Queued all EBMS samples from 2023 (by date_start) to have their locations re-indexed. 
6. On completion of the previous step, queued all EBMS occurrences from 2023 to have their locations re-indexed.

This seems to have resolved all problems with coastal records not appearing to be in the country. It is recognised that this test now means the land borders between Spain and neighbouring countries are inconsistent so a record could potentially appear in both countries due to an overlap in borders. Therefore, a complete roll out of updated country borders is desirable.

**Comment by JimBacon** [comment]
Maps for the following countries have been updated: Belgium, Croatia, Denmark, Estonia, Finland, France, Germany, Greece, Italy, Latvia, Lithuania, Netherlands, Norway, Poland, Portugal, Russian Federation, Slovenia, Sweden.

The maps were taken from the same source as above. With the following modifications:

- Portugal was merged with Madeira and the Azores, consistent with its previous definition.
- France was merged with Réunion, Mayotte, French Guiana, Martinique, Guadeloupe, and the Collectivity of Saint Martin, consistent with its previous definition.
- Some hand editing of the borders of Belgium/Netherlands, and Croatia with Slovenia, Serbia, & Bosnia was done where it significantly deviated from the 1:1million boundaries in the 2016 Eurostat maps.

The following two queries were used to reindex EBMS records from 2023.

```sql
insert into work_queue(task, entity, record_id, params, cost_estimate, priority, created_on)
select 'task_spatial_index_builder_sample', 'sample', id, null, 100, 2, now()
from cache_samples_functional 
where website_id = 118 -- EBMS
and date_part('year', date_start) = '2023' 
order by id;
```

```sql
insert into work_queue(task, entity, record_id, params, cost_estimate, priority, created_on)
select 'task_spatial_index_builder_occurrence', 'occurrence', id, null, 100, 2, now()
from cache_occurrences_functional 
where website_id = 118 -- EBMS
and date_part('year', date_start) = '2023' 
order by id;
```

**Comment by JimBacon** [comment]
Until all the inland country boundaries are updated, which is not such a high priority as fixing the coastal boundaries, be prepared for occasional anomalies due to the mixing of maps from different sources. There may be holes, where boundaries do not meet, or overlaps. The consequence of this would be records either not being assigned to a country (hole) or to two countries (overlap). My current impression is that the best map would be a combination of the 1:1million Eurostat internal boundaries combined with the EEZ marine boundaries.

**Comment by DavidRoy** [comment]
@JimBacon I think this can be closed for now? Specific issue raised as needed?

**Comment by JimBacon** [comment]
@DavidRoy I will 'Close as not planned'
Hopefully someone will remember these details if we need to address this again.
